clients/go/pkg/zbc/client.go

```diff
@@ -83,72 +85,94 @@ func (e Error) Error() string {
 	return string(e)
 }
 
+func withStopOnPermanentError(shouldRetryRequest func(ctx context.Context, err error) bool) func(ctx context.Context, err error) bool {
+	return func(ctx context.Context, err error) bool {
+		switch {
+		case errors.Is(err, context.Canceled):
+			return false
+		case errors.Is(err, context.DeadlineExceeded):
+			return false
+		}
+
+		switch status.Code(err) {
+		case codes.Canceled, codes.DeadlineExceeded, codes.InvalidArgument, codes.Unimplemented:
```

Comment:
I would add as well:

- `FAILED_PRECONDITION`: we return this when there's the request was well formed, but is not processable. Typically doesn't make sense to retry unless you modify the state with some other command.
- `ALREADY_EXISTS`: we return this in a few cases (e.g. unique message IDs), doesn't make sense to retry automatically.
- `PERMISSION_DENIED`: we return this when you have the wrong tenant. Also doesn't make sense to retry without first changing user, for example, which requires re-configuring the client.

As we do sometimes return them, and both are not recoverable without a change of the system state (typically via another command, not this one). 

I suppose `NOT_FOUND` also doesn't make much sense to retry automatically? I'm not sure on that one, since it could happen that you deploy a process, and try to start a PI for it before it's been distributed, so... :thinking: